### DISCUSSION TITLE
How to cancel a watcher request?

### DISCUSSION THREAD

**Discussion by goenning** [kube-rs/kube#1004] [Q&A]
Hi all,

I have a watcher running on a tokio task, something like this:

```
let handle = spawn(async move {
    let w = watcher(api, params).backoff(default_backoff());

    pin_mut!(w);

    loop {
        let event = w.try_next().await;

        match event {
            Ok(Some(Event::Applied(mut e))) => {
                ...
```

Eventually I want to cancel the watcher so I just do a `handle.abort()`. I thought this would be enough to cancel the underlaying request, but it doesn't. While using proxyman, the watch HTTP request is still active, and it stays active until the next event arrives or it times out. 

Is there a way to cancel the request as well? It's a minor optimization, but I'd prefer not having unnecessary open connections.

Thank you!

**Comment by nightkr** [comment]
That should be sufficient to cancel, this sounds like a bug on our side.

**Reply by nightkr** [reply]
Created an issue: #1005

**Comment by nightkr** [comment]
Nevermind.. after creating a more representative repro (https://gist.github.com/teozkr/cc553060e77027ebc82a10388e8f7505) I am unable to recreate this problem. As long as the watcher is owned by the tokio task (such as when it's created inside of the `async` block) then `JoinHandle::abort` should abort the watch request correctly.

**Comment by goenning** [comment]
Thanks for looking into it. In your repro, how do you know the request was aborted? 

Here's a video of what I'm seeing with your repro.
1. because kube-rs doesn't support http proxy, I start a proxy with kubectl `http_proxy=http://127.0.01:9090 https_proxy=http://127.0.01:9090 kubectl proxy --insecure-skip-tls-verify=true`
2. I then change the current context cluster to `http://127.0.0.1:8001` (kubectl proxy), so that kube-rs -> kubectl proxy -> proxyman -> apiserver
3. Then start your example

https://user-images.githubusercontent.com/94755/189356657-2afa97e0-f619-49b2-92db-315863049bcc.mp4

What is super weird here is that even after the program exited completely, the underlying connection still shows as active on proxyman, which to my understand should not be possible.

I wonder if kube-rs is actually aborting as expected, but the kubectl proxy doesn't, so it keeps showing as active on proxyman. I wish I could proxy kube-rs directly to proxyman though, that would simplify this.

Am I doing something wrong?

**Reply by nightkr** [reply]
I used Wireshark for the communication between `kubectl proxy` and the repro program to confirm that the last communication happened when dropping the watcher. You can also set `RUST_LOG=trace` to get more information from `hyper` about the state of the HTTP connection.
